# Ejercicio 7

> Por: Mapachana

**Toma tu número n de la lista publicada para el ejercicio 3. Sea d elprimer elemento de la sucesión 5, -7, 9, -11, 13, ... que satisface que el símbolo deJacobi es (d|n) = -1**

In [1]:
n = 36580545945776718558633000960211
n

36580545945776718558633000960211

In [2]:
def jacobi(a, n):
        #assert(n > a > 0 and n%2 == 1)
        t = 1
        while a != 0:
            while a % 2 == 0:
                a /= 2
                r = n % 8
                if r == 3 or r == 5:
                    t = -t
            a, n = n, a
            if a % 4 == n % 4 == 3:
                t = -t
            a %= n
        if n == 1:
            return t
        else:
            return 0

In [3]:
num_base = 5
exp = 2

while True:
    num_suc = pow(-1, exp)*num_base
    print(num_suc)
    
    if jacobi(num_suc, n) == -1:
        print(num_suc)
        break
    
    if exp >= 10:
        break
    
    num_base += 2
    exp += 1
    
print(jacobi(5, n))
print(jacobi(-7, n))
print(jacobi(9, n))
print(jacobi(-11, n))
print(jacobi(13, n))

5
-7
9
-11
13
13
1
0
1
0
-1


In [4]:
d = -7
d

-7

Luego el primer número de la sucesión que cumple $\left( \frac{d}{n} \right) = -1$ es 13, y por tanto $d=-7$.

In [5]:
import math

**i) Con P = 1, Q = (1-d)/4, define el e.c. $\alpha$ y sus sucesiones de Lucas asociadas.**

In [6]:
P = 1
Q = (1-d)/4

#alfa = (P+math.sqrt(pow(P,2)-4*Q))/2
#print(alfa)
print(pow(P,2)-4*Q)
print(Q)
#print((1+math.sqrt(13))/2)

-7.0
2.0


Sabemos que $\alpha = \frac{P+\sqrt{P^2-4Q}}{2}$, luego tenemos que $\alpha = \frac{1+\sqrt{-7}}{2}$

Además, sabemos que $\alpha$ es un entero cuadrático por ser solución de $x^2-x+2=0 \in \mathbb{Z}[x]$.

Entonces, se tiene $\alpha^n = \alpha^{n-1}-2\alpha^{n-2}$.

Por tanto si llamamos $\alpha^{i}=\frac{V_i}{2}+\frac{U_i}{2}\sqrt{-7}$ se tiene 
$$U_0=0, \quad V_0=2$$
$$U_1=1, \quad V_1=1$$


La sucesión de Lucas asociada es:

$$V_n=V_{n-1}-2V_{n-2}$$
$$U_n=U_{n-1}-2U_{n-2}$$

**ii) Si n primo ¿Que debería de pasarle a $V_r,U_r$,módulo n? ¿Y a $V_{r/2} ,U_{r/2}$? Calcula los términos $V_{r/2} ,U_{r/2},V_r ,U_r$, módulo n, de las sucesiones de Lucas. ¿ Tu n verifica el TPF para el entero cuadrático $\alpha$ ?**

Si $n$ fuera primo por la tercera versión del TPF para e.c. se tiene que, con $\Delta = P^2-4Q=13$:

$$U_r = U_{n-\frac{\Delta}{n}}=U_{n+1}\equiv 0\mod n$$
$$V_r = V_{n-\frac{\Delta}{n}}=V_{n+1}\equiv 2Q\mod n\equiv 4\mod n$$

Donde hemos usado que $\left( \frac{\Delta}{P}\right) = -1$

Además, conocemos las propiedades:

$$U_{2k}=U_kV_k$$
$$V_{2k}=V_k^2-2Q^k$$

Sabemos que $r=n+1$, y, con $k=\frac{r}{2}$:

$$U_r=U_{\frac{r}{2}}V_{\frac{r}{2}}$$
$$V_r=V_{\frac{r}{2}}^2-2Q^{\frac{r}{2}}$$

$$U_r=U_{\frac{r}{2}}V_{\frac{r}{2}}= 0 \mod n$$
Luego $U_{\frac{r}{2}}$ o $V_{\frac{r}{2}}$ son 0 o múltiplos de $n$.

Por otro lado, 

$$V_{\frac{r}{2}}^2=4+2(2)^{\frac{r}{2}}=36580545945776718558633000960211 \mod n \equiv 0 \mod n$$.


In [7]:
delta = pow(P,2)-4*Q
print(jacobi(delta, n))
print("AKJHJAKS")

print(2*Q)
print(Q)


aux = pow(int(Q), (n+1)//2, n)
aux = 4+pow(2*aux, 1, n)
print(aux)
print(aux-n)

0
AKJHJAKS
4.0
2.0
36580545945776718558633000960211
0


In [8]:
def calcular_digitos(b):
    digitos = []
    
    while b != 1:
        resto = pow(b,1,2)
        digitos.append(int(resto))
        b = (b-resto)//2
    
    digitos.append(1)
    
    digitos = list(reversed(digitos))
        
    return digitos

In [9]:
calcular_digitos(13)

[1, 1, 0, 1]

In [10]:
def suc_lucas(n, p, q):
    r = int(n+1) # solo porque en este caso el jacobi es -1
    digitos = calcular_digitos(r)
    #print(r)
    
    U = [-1, -1]
    U[0], U[1] = 0, 1
    
    U_nuevo = [-1, -1]
    V = -1
    
    V = 2*U[1]-p*U[0]
    
    j = 0
    digit = 0
    #print(digitos)
    
    print("U[{}] = {}, \nV[{}] = {}, \n\nU[{}] = {}\n".format(2*j, U[0], 2*j, V, 2*j+1, U[1]))
    while digit < len(digitos):
        #print(digitos[digit])
        
        if digitos[digit] == 0:
            aux = 2*U[0]*U[1]-p*pow(U[0], 2)
            aux = pow(aux, 1, n)
            U_nuevo[0] = aux
            
            aux = pow(U[1], 2)-q*pow(U[0], 2)
            aux = pow(aux, 1, n)
            U_nuevo[1] = aux
            
            V = pow(2*U_nuevo[1]-p*U_nuevo[0],1,n)
            
            print("U[{}] = {}, \nV[{}] = {}, \n\nU[{}] = {}\n".format(2*j, U_nuevo[0], 2*j, V, 2*j+1, U_nuevo[1]))
            
            j = 2*j
        else:
            aux = pow(U[1], 2)-q*pow(U[0], 2)
            aux = pow(aux, 1, n)
            U_nuevo[0] = aux
            
            aux = p*pow(U[1],2)-2*q*U[0]*U[1]
            aux = pow(aux, 1, n)
            U_nuevo[1] = aux
            
            V = pow(2*U_nuevo[1]-p*U_nuevo[0],1,n)
            
            print("U[{}] = {}, \nV[{}] = {}, \n\nU[{}] = {}\n".format(2*j+1, U_nuevo[0], 2*j+1, V, 2*j+2, U_nuevo[1]))
            
            j = 2*j+1
            
        if 2*j == r//2 or 2*j+1 == r//2:
            print("\n\n")
            print("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")
             
        U[0], U[1] = U_nuevo[0], U_nuevo[1]
        
        digit += 1

In [11]:
suc_lucas(11,3,-1)
print("\n")
suc_lucas(5914266590884932403584601505279, 1, 2)
#print(calcular_digitos(5914266590884932403584601505279))

U[0] = 0, 
V[0] = 2, 

U[1] = 1

U[1] = 1, 
V[1] = 3, 

U[2] = 3

U[3] = 10, 
V[3] = 3, 

U[4] = 0




AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
U[6] = 8, 
V[6] = 0, 

U[7] = 1

U[12] = 0, 
V[12] = 9, 

U[13] = 10



U[0] = 0, 
V[0] = 2, 

U[1] = 1

U[1] = 1, 
V[1] = 1, 

U[2] = 1

U[2] = 1, 
V[2] = 5914266590884932403584601505276, 

U[3] = 5914266590884932403584601505278

U[4] = 5914266590884932403584601505276, 
V[4] = 1, 

U[5] = 5914266590884932403584601505278

U[9] = 5914266590884932403584601505262, 
V[9] = 5914266590884932403584601505274, 

U[10] = 5914266590884932403584601505268

U[18] = 85, 
V[18] = 5914266590884932403584601504280, 

U[19] = 5914266590884932403584601504822

U[37] = 194399, 
V[37] = 534059, 

U[38] = 364229

U[74] = 103820535541, 
V[74] = 10341108537, 

U[75] = 57080822039

U[149] = 5914266572585745448192501297438, 
V[149] = 5914266568291267742596713243370, 

U[150] = 5914266570438506595394607270404

U[298] = 4936250694530866754414472699981, 
V[

In [12]:
suc_lucas(n, P, int(Q))

U[0] = 0, 
V[0] = 2, 

U[1] = 1

U[1] = 1, 
V[1] = 1, 

U[2] = 1

U[3] = 36580545945776718558633000960210, 
V[3] = 36580545945776718558633000960206, 

U[4] = 36580545945776718558633000960208

U[7] = 7, 
V[7] = 36580545945776718558633000960198, 

U[8] = 36580545945776718558633000960208

U[14] = 36580545945776718558633000960120, 
V[14] = 36580545945776718558633000960124, 

U[15] = 36580545945776718558633000960122

U[28] = 7917, 
V[28] = 36580545945776718558633000935012, 

U[29] = 36580545945776718558633000951570

U[57] = 36580545945776718558632950269314, 
V[57] = 747311035, 

U[58] = 348310069

U[115] = 116180770089455543, 
V[115] = 267708236898049123, 

U[116] = 191944503493752333

U[230] = 9085068195533530414279122459439, 
V[230] = 4080701587084059651838353133425, 

U[231] = 6582884891308795033058737796432

U[461] = 5959030677211170876601815288653, 
V[461] = 12231053807303017165974138820499, 

U[462] = 9095042242257094021287977054576

U[923] = 4246659038619848551495442516732, 
V[923] =

De donde tenemos que $U_r \equiv 0\mod n$ y $V_r\equiv 4\mod n$. Además tenemos que $U_{\frac{r}{2}}\equiv 21008262978350532486723653312533\mod n$ y $V_{\frac{r}{2}}\equiv 0\mod n$.

Por tanto, tenemos que $n$ verifica el TPF para el e.c. $\alpha$.

In [13]:

18290272972888359279316500480106 - 18290272972888358803181567737856

476134932742250

**iii) Factoriza r = n+1 y para cada factor primo p suyo, calcula $U_{r/p}$ . ¿ Cuál es el rango de Lucas w(n) ?. ¿ Qué deduces sobre la primalidad de tu n ?**

In [58]:
r = n+1
r

36580545945776718558633000960212

In [59]:
def f(x):
    return pow(x,2)+1

def rho_polar(n, x0, t0):
    x = x0
    y = x
    t = t0
    i = 0
    g= '-'
    print("Paso: {}, x: {}, y: {}, g: {}".format(i,x,y,g))
    while i < t:
        i = i+1
        x = pow(f(x),1,n)
        y = pow(f(f(y)),1,n)
        g = math.gcd(x-y, n)
        print("Paso: {}, x: {}, y: {}, g: {}".format(i,x,y,g))
        if 1 < g and g < n:
            print("g: {}".format(g))
            print("Iteraciones necesarias {}".format(i))
            
            return g, i
    
    print("No hay divisores con t iteraciones")
    return "No hay divisores con t iteraciones"

In [60]:
m = r
i = 0
factores = []

while pow(m, 1, 2) == 0:
    i += 1
    m = m//2
    factores.append(2)


print(i)
print(m)
print(factores)

2
9145136486444179639658250240053
[2, 2]


In [61]:
itotal = 0
g, i = rho_polar(m, 1, 100)
factores.append(g)
m = m//g
itotal = itotal + i

#factores.append(m)
print(factores)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 677, g: 1
Paso: 3, x: 26, y: 210066388901, g: 1
Paso: 4, x: 677, y: 1769955215995455558797896924155, g: 19
g: 19
Iteraciones necesarias 4
[2, 2, 19]
4


In [62]:
g, i = rho_polar(m, 1, 100)
factores.append(g)
m = m//g
itotal = itotal + i

#factores.append(m)
print(factores)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 677, g: 1
Paso: 3, x: 26, y: 210066388901, g: 1
Paso: 4, x: 677, y: 325986297083216668325541623094, g: 1
Paso: 5, x: 458330, y: 127410829911494166241278775365, g: 1
Paso: 6, x: 210066388901, y: 68310586424312394394477383158, g: 53
g: 53
Iteraciones necesarias 6
[2, 2, 19, 53]
10


In [63]:
g, i = rho_polar(m, 1, 1000)
factores.append(g)
m = m//g
itotal = itotal + i

factores.append(m)
print(m)
print(factores)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 677, g: 1
Paso: 3, x: 26, y: 210066388901, g: 1
Paso: 4, x: 677, y: 8131503611969113819048317829, g: 1
Paso: 5, x: 458330, y: 268912522995144438681453259, g: 1
Paso: 6, x: 210066388901, y: 4739627730062883493178722105, g: 1
Paso: 7, x: 44127887745906175987802, y: 2877805303710922471854403996, g: 1
Paso: 8, x: 8131503611969113819048317829, y: 844905932928462989573800276, g: 1
Paso: 9, x: 4808195559081928123140273788, y: 8036760750550991519670821456, g: 1
Paso: 10, x: 268912522995144438681453259, y: 612178491547928336261128025, g: 1
Paso: 11, x: 1441626329765841981061538451, y: 2750291903314505890023313716, g: 1
Paso: 12, x: 4739627730062883493178722105, y: 7646807715136505100599503640, g: 1
Paso: 13, x: 6053686783633123615568617369, y: 3976114632546841138731907572, g: 1
Paso: 14, x: 2877805303710922471854403996, y: 3291392848644434084727765554, g: 1
Paso: 15, x: 378439376000717070129403836, y: 863675745714514403263550

Estos son todos los factores, los dos ultimos son primos pero falta ver por que es primo pero falta ver por que

In [20]:
pow(11, 43117-1, 43117)

1

In [21]:
print(factores)
2*2*19*53*43117*210626099398147601381287

[2, 2, 19, 53, 43117, 210626099398147601381287]


36580545945776718558633000960212

In [22]:
def lucas_lehmer(n, primos, num):
    fermat = False
    for a in range (1, num):
        if pow(a, n-1, n) == 1:
            fermat = True
            print(a)
            print("test fermat: {}".format(pow(a, n-1, n)))
            for num in primos:
                print("Para num: {}".format(pow(a, (n-1)//num, n)))

### Comprobando primalidad de 43117

Para ello, comenzamos descomponiendo con rho de Polard 43117-1:

In [65]:
m = 43117-1

i = 0
primos = []

while pow(m, 1, 2) == 0:
    i += 1
    m = m//2

primos.append(2)

print(i)
print(m)
print(primos)

2
10779
[2]


In [66]:
itotal = 0
g, i = rho_polar(m, 1, 1000)
primos.append(g)
m = m//g
itotal = itotal + i

primos.append(m)
print(m)
print(primos)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 3
g: 3
Iteraciones necesarias 1
3593
[2, 3, 3593]
1


In [25]:
2*2*3*3593 == 43117-1

True

In [26]:
lucas_lehmer(43117, primos, 10)

1
test fermat: 1
Para num: 1
Para num: 1
Para num: 1
2
test fermat: 1
Para num: 43116
Para num: 9558
Para num: 4096
3
test fermat: 1
Para num: 1
Para num: 1
Para num: 14037
4
test fermat: 1
Para num: 1
Para num: 33558
Para num: 4703
5
test fermat: 1
Para num: 43116
Para num: 1
Para num: 12171
6
test fermat: 1
Para num: 43116
Para num: 9558
Para num: 20591
7
test fermat: 1
Para num: 1
Para num: 33558
Para num: 40329
8
test fermat: 1
Para num: 43116
Para num: 1
Para num: 33306
9
test fermat: 1
Para num: 1
Para num: 1
Para num: 35796


In [27]:
print(pow(2, (43117-1)//2, 43117))
print(pow(2, (43117-1)//3, 43117))
print(pow(2, (43117-1)//3593, 43117))

43116
9558
4096


Tenemos que $a=2$ es un elemento primitivo para $43117$ porque $2^{43117-1}\equiv 1\mod 43117$ y $2^{\frac{43117-1}{p}}\not\equiv 1\mod 43117 $ para $p\in\{2,3,3593\}$ pues:

$$2^{\frac{43117-1}{2}}\equiv 43116\mod 43117$$
$$2^{\frac{43117-1}{3}}\equiv 9558\mod 43117$$
$$2^{\frac{43117-1}{3593}}\equiv 4096\mod 43117$$

### Comprobando primalidad de 210626099398147601381287

Llamamos $m=210626099398147601381287$

Para ello, comenzamos descomponiendo con rho de Polard m-1:

In [73]:
m = 210626099398147601381287-1

i = 0
primos = []

while pow(m, 1, 2) == 0:
    i += 1
    m = m//2

primos.append(2)

print(i)
print(m)
print(primos)

1
105313049699073800690643
[2]


In [74]:
itotal = 0
g, i = rho_polar(m, 1, 1000)
primos.append(g)
m = m//g
itotal = itotal + i

#primos.append(m)
print(m)
print(primos)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 3
g: 3
Iteraciones necesarias 1
35104349899691266896881
[2, 3]
1


In [75]:
g, i = rho_polar(m, 1, 1000)
primos.append(g)
m = m//g
itotal = itotal + i

#primos.append(m)
print(m)
print(primos)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 677, g: 1
Paso: 3, x: 26, y: 210066388901, g: 1
Paso: 4, x: 677, y: 11455057248607197263599, g: 113
g: 113
Iteraciones necesarias 4
310657963714082007937
[2, 3, 113]
5


In [76]:
g, i = rho_polar(m, 1, 1000)
primos.append(g)
m = m//g
itotal = itotal + i

#primos.append(m)
print(m)
print(primos)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 677, g: 1
Paso: 3, x: 26, y: 210066388901, g: 1
Paso: 4, x: 677, y: 271370554900244977867, g: 1
Paso: 5, x: 458330, y: 243637929663945897992, g: 1
Paso: 6, x: 210066388901, y: 61390790324343477542, g: 17
g: 17
Iteraciones necesarias 6
18273997865534235761
[2, 3, 113, 17]
11


In [32]:
g, i = rho_polar(m, 1, 1000)
primos.append(g)
m = m//g
itotal = itotal + i

#primos.append(m)
print(m)
print(primos)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 677, g: 1
Paso: 3, x: 26, y: 210066388901, g: 1
Paso: 4, x: 677, y: 15534584782765677213, g: 1
Paso: 5, x: 458330, y: 6075957412000833099, g: 1
Paso: 6, x: 210066388901, y: 6568796727740770259, g: 1
Paso: 7, x: 14456898506530860748, y: 1871839733591523871, g: 1
Paso: 8, x: 15534584782765677213, y: 16428899933174250490, g: 1
Paso: 9, x: 12607330261578787802, y: 3417119749382987037, g: 1
Paso: 10, x: 6075957412000833099, y: 10463363848283700167, g: 1
Paso: 11, x: 2199084128639880363, y: 16194572544503398000, g: 1
Paso: 12, x: 6568796727740770259, y: 4421289643826340042, g: 1
Paso: 13, x: 10780582564206380224, y: 11781826174035754604, g: 1
Paso: 14, x: 1871839733591523871, y: 6209601081096985630, g: 1
Paso: 15, x: 17258480816574362426, y: 13105134950878194682, g: 1
Paso: 16, x: 16428899933174250490, y: 8606959312254178812, g: 1
Paso: 17, x: 12718437182065319708, y: 9886235276795346979, g: 1
Paso: 18, x: 3417119749382987

FALTA COMPROBAR 25373

In [77]:
g, i = rho_polar(m, 1, 1000)
primos.append(g)
m = m//g
itotal = itotal + i

#primos.append(m)
print(m)
print(primos)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 677, g: 1
Paso: 3, x: 26, y: 210066388901, g: 1
Paso: 4, x: 677, y: 15534584782765677213, g: 1
Paso: 5, x: 458330, y: 6075957412000833099, g: 1
Paso: 6, x: 210066388901, y: 6568796727740770259, g: 1
Paso: 7, x: 14456898506530860748, y: 1871839733591523871, g: 1
Paso: 8, x: 15534584782765677213, y: 16428899933174250490, g: 1
Paso: 9, x: 12607330261578787802, y: 3417119749382987037, g: 1
Paso: 10, x: 6075957412000833099, y: 10463363848283700167, g: 1
Paso: 11, x: 2199084128639880363, y: 16194572544503398000, g: 1
Paso: 12, x: 6568796727740770259, y: 4421289643826340042, g: 1
Paso: 13, x: 10780582564206380224, y: 11781826174035754604, g: 1
Paso: 14, x: 1871839733591523871, y: 6209601081096985630, g: 1
Paso: 15, x: 17258480816574362426, y: 13105134950878194682, g: 1
Paso: 16, x: 16428899933174250490, y: 8606959312254178812, g: 1
Paso: 17, x: 12718437182065319708, y: 9886235276795346979, g: 1
Paso: 18, x: 3417119749382987

In [34]:
g, i = rho_polar(m, 1, 1000)
primos.append(g)
m = m//g
itotal = itotal + i

primos.append(m)
print(m)
print(primos)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 677, g: 1
Paso: 3, x: 26, y: 210066388901, g: 1
Paso: 4, x: 677, y: 185270304257, g: 1
Paso: 5, x: 458330, y: 497898891190, g: 1
Paso: 6, x: 210066388901, y: 5923262369, g: 1
Paso: 7, x: 166547059314, y: 758564388970, g: 1
Paso: 8, x: 185270304257, y: 752575103895, g: 1
Paso: 9, x: 258546563355, y: 529949035159, g: 1
Paso: 10, x: 497898891190, y: 838048329332, g: 1
Paso: 11, x: 598172135496, y: 446879879228, g: 1
Paso: 12, x: 5923262369, y: 67853697551, g: 1
Paso: 13, x: 28102662610, y: 318842701668, g: 1
Paso: 14, x: 758564388970, y: 687400713657, g: 1
Paso: 15, x: 858939208099, y: 279361155417, g: 1
Paso: 16, x: 752575103895, y: 288361175417, g: 1
Paso: 17, x: 610474637070, y: 730735073390, g: 1
Paso: 18, x: 529949035159, y: 631885725007, g: 1
Paso: 19, x: 449058523368, y: 827895846876, g: 1
Paso: 20, x: 838048329332, y: 857870536977, g: 1
Paso: 21, x: 581557572234, y: 291493040986, g: 1
Paso: 22, x: 446879879228, 

Paso: 291, x: 554048222491, y: 248292245927, g: 1
Paso: 292, x: 952136680117, y: 925847108438, g: 1
Paso: 293, x: 555868555019, y: 916656221252, g: 1
Paso: 294, x: 838859169084, y: 618912527099, g: 1
Paso: 295, x: 719596386406, y: 410066760328, g: 1
Paso: 296, x: 572600438215, y: 854106567478, g: 1
Paso: 297, x: 965682324422, y: 840968804441, g: 1
Paso: 298, x: 710830003221, y: 790522138440, g: 1
Paso: 299, x: 446613949619, y: 464209609631, g: 1
Paso: 300, x: 594256451733, y: 604091713549, g: 1
Paso: 301, x: 406462896698, y: 885365505698, g: 1
Paso: 302, x: 164634958138, y: 427038661580, g: 1
Paso: 303, x: 75503413788, y: 787967768680, g: 1
Paso: 304, x: 44194817008, y: 171556781271, g: 1
Paso: 305, x: 574121430998, y: 413939876244, g: 1
Paso: 306, x: 671546814072, y: 384117165083, g: 1
Paso: 307, x: 53306900650, y: 760141710492, g: 1
Paso: 308, x: 412143819176, y: 631270921819, g: 1
Paso: 309, x: 805856041257, y: 916216559578, g: 1
Paso: 310, x: 467238731361, y: 825508745575, g: 1
Pas

In [35]:
print(primos)
lucas_lehmer(210626099398147601381287, primos, 10)

[2, 3, 113, 17, 25373, 733, 463867, 2118187]
1
test fermat: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 1
2
test fermat: 1
Para num: 1
Para num: 1
Para num: 145657945091506546457488
Para num: 192364560553782260657209
Para num: 7748558939617545301762
Para num: 171185533527692094621716
Para num: 135754665705968357010284
Para num: 31574681796135134908225
3
test fermat: 1
Para num: 210626099398147601381286
Para num: 1
Para num: 145232060748561743442937
Para num: 159211811116237223074325
Para num: 54725823724448612498237
Para num: 191210551915139078613713
Para num: 23327454618920415194213
Para num: 88903178243381857375732
4
test fermat: 1
Para num: 1
Para num: 1
Para num: 155696556272473449748877
Para num: 118578062198627107902407
Para num: 30623794958215767949216
Para num: 184622898293135382921935
Para num: 147475185621693620913810
Para num: 81353323058220189225649
5
test fermat: 1
Para num: 210626099398147601381286
Para num: 201052306676

FALTA COMPROBAR 463867 Y 2118187

#### Comprobando primalidad de 25373

Llamamos $m=25373$

Para ello, comenzamos descomponiendo con rho de Polard m-1:

In [78]:
m = 25373-1

i = 0
primos = []

while pow(m, 1, 2) == 0:
    i += 1
    m = m//2

primos.append(2)
primos.append(m)
print(i)
print(m)
print(primos)

2
6343
[2, 6343]


In [37]:
2*2*6343 == 25373-1

True

6343 es primo por la lista, luego aplicamos Lucas Lehmer

In [38]:
lucas_lehmer(25373, primos,40)

1
test fermat: 1
Para num: 1
Para num: 1
2
test fermat: 1
Para num: 25372
Para num: 16
3
test fermat: 1
Para num: 25372
Para num: 81
4
test fermat: 1
Para num: 1
Para num: 256
5
test fermat: 1
Para num: 25372
Para num: 625
6
test fermat: 1
Para num: 1
Para num: 1296
7
test fermat: 1
Para num: 25372
Para num: 2401
8
test fermat: 1
Para num: 25372
Para num: 4096
9
test fermat: 1
Para num: 1
Para num: 6561
10
test fermat: 1
Para num: 1
Para num: 10000
11
test fermat: 1
Para num: 25372
Para num: 14641
12
test fermat: 1
Para num: 25372
Para num: 20736
13
test fermat: 1
Para num: 1
Para num: 3188
14
test fermat: 1
Para num: 1
Para num: 13043
15
test fermat: 1
Para num: 1
Para num: 25252
16
test fermat: 1
Para num: 1
Para num: 14790
17
test fermat: 1
Para num: 1
Para num: 7402
18
test fermat: 1
Para num: 25372
Para num: 3484
19
test fermat: 1
Para num: 25372
Para num: 3456
20
test fermat: 1
Para num: 25372
Para num: 7762
21
test fermat: 1
Para num: 1
Para num: 16870
22
test fermat: 1
Para num

El 2 es elemento primitivo para 25373, luego es primo.

### Comprobando primalidad de 463867

Llamamos $m=463867$

Para ello, comenzamos descomponiendo con rho de Polard m-1:

In [79]:
m = 463867-1

i = 0
primos = []

while pow(m, 1, 2) == 0:
    i += 1
    m = m//2

primos.append(2)
#primos.append(m)
print(i)
print(m)
print(primos)

1
231933
[2]


In [80]:
itotal = 0
g, i = rho_polar(m, 1, 100)
primos.append(g)
m = m//g
itotal = itotal + i

#primos.append(m)
print(primos)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 3
g: 3
Iteraciones necesarias 1
[2, 3]
1


In [81]:
g, i = rho_polar(m, 1, 100)
primos.append(g)
m = m//g
itotal = itotal + i

primos.append(m)
print(m)
print(primos)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 677, g: 1
Paso: 3, x: 26, y: 32141, g: 1
Paso: 4, x: 677, y: 2406, g: 247
g: 247
Iteraciones necesarias 4
313
[2, 3, 247, 313]
5


In [82]:
m = 247
primos = [2, 3, 313]
g, i = rho_polar(m, 1, 100)
primos.append(g)
m = m//g
itotal = itotal + i

primos.append(m)
print(primos)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 183, g: 1
Paso: 3, x: 26, y: 31, g: 1
Paso: 4, x: 183, y: 183, g: 247
Paso: 5, x: 145, y: 31, g: 19
g: 19
Iteraciones necesarias 5
[2, 3, 313, 19, 13]
10


In [43]:
2*3*313*19*13 == 463867-1

True

In [44]:
lucas_lehmer(463867, primos, 20)

1
test fermat: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 1
2
test fermat: 1
Para num: 463866
Para num: 1
Para num: 332498
Para num: 193762
Para num: 292075
3
test fermat: 1
Para num: 463866
Para num: 47777
Para num: 253250
Para num: 361670
Para num: 1
4
test fermat: 1
Para num: 1
Para num: 1
Para num: 106293
Para num: 173132
Para num: 344990
5
test fermat: 1
Para num: 463866
Para num: 1
Para num: 232740
Para num: 413484
Para num: 74396
6
test fermat: 1
Para num: 1
Para num: 47777
Para num: 269724
Para num: 123249
Para num: 292075
7
test fermat: 1
Para num: 1
Para num: 416089
Para num: 422129
Para num: 67751
Para num: 444295
8
test fermat: 1
Para num: 463866
Para num: 1
Para num: 183184
Para num: 5011
Para num: 372909
9
test fermat: 1
Para num: 1
Para num: 416089
Para num: 383346
Para num: 261304
Para num: 1
10
test fermat: 1
Para num: 1
Para num: 1
Para num: 44511
Para num: 234036
Para num: 289819
11
test fermat: 1
Para num: 1
Para num: 416089
Para num: 46951
Para num

12 es elemento primitivo luego es primo

### Comprobando primalidad de 2118187

Llamamos $m=2118187$

Para ello, comenzamos descomponiendo con rho de Polard m-1:

In [84]:
m = 2118187-1

i = 0
primos = []

while pow(m, 1, 2) == 0:
    i += 1
    m = m//2

primos.append(2)
#primos.append(m)
print(i)
print(m)
print(primos)

1
1059093
[2]


In [85]:
itotal = 0
g, i = rho_polar(m, 1, 100)
primos.append(g)
m = m//g
itotal = itotal + i

#primos.append(m)
print(primos)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 3
g: 3
Iteraciones necesarias 1
[2, 3]
1


In [86]:
g, i = rho_polar(m, 1, 100)
primos.append(g)
m = m//g
itotal = itotal + i

#primos.append(m)
print(primos)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 3
g: 3
Iteraciones necesarias 1
[2, 3, 3]
2


In [87]:
g, i = rho_polar(m, 1, 100)
primos.append(g)
m = m//g
itotal = itotal + i

primos.append(m)
print(m)
print(primos)
print(itotal)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 677, g: 7
g: 7
Iteraciones necesarias 2
16811
[2, 3, 3, 7, 16811]
4


In [49]:
2*3*3*7*16811 == 2118187-1

True

In [50]:
lucas_lehmer(2118187, primos, 20)

1
test fermat: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 1
2
test fermat: 1
Para num: 2118186
Para num: 1491183
Para num: 1491183
Para num: 1355468
Para num: 1526343
3
test fermat: 1
Para num: 2118186
Para num: 1491183
Para num: 1491183
Para num: 1
Para num: 1820979
4
test fermat: 1
Para num: 1
Para num: 627003
Para num: 627003
Para num: 1395281
Para num: 1090707
5
test fermat: 1
Para num: 2118186
Para num: 1491183
Para num: 1491183
Para num: 1481444
Para num: 1971333
6
test fermat: 1
Para num: 1
Para num: 627003
Para num: 627003
Para num: 1355468
Para num: 168511
7
test fermat: 1
Para num: 2118186
Para num: 1
Para num: 1
Para num: 1
Para num: 1152592
8
test fermat: 1
Para num: 2118186
Para num: 1
Para num: 1
Para num: 1592566
Para num: 1803664
9
test fermat: 1
Para num: 1
Para num: 627003
Para num: 627003
Para num: 1
Para num: 2079177
10
test fermat: 1
Para num: 1
Para num: 627003
Para num: 627003
Para num: 187044
Para num: 1209792
11
test fermat: 1
Para num: 2118186

$16811=2*5*41*41$, donde estos numeros son primos. Luegopara la descomposicion en primos de lucas lehmer:

In [51]:
(16811-1)//2

8405

In [52]:
8405//5

1681

In [53]:
1681/41

41.0

In [54]:
primos = [2, 5, 41]

In [55]:
lucas_lehmer(16811, primos, 10)

1
test fermat: 1
Para num: 1
Para num: 1
Para num: 1
2
test fermat: 1
Para num: 16810
Para num: 1
Para num: 15561
3
test fermat: 1
Para num: 1
Para num: 1
Para num: 4556
4
test fermat: 1
Para num: 1
Para num: 1
Para num: 15888
5
test fermat: 1
Para num: 1
Para num: 1
Para num: 12976
6
test fermat: 1
Para num: 16810
Para num: 1
Para num: 3929
7
test fermat: 1
Para num: 16810
Para num: 2954
Para num: 11826
8
test fermat: 1
Para num: 16810
Para num: 1
Para num: 10602
9
test fermat: 1
Para num: 1
Para num: 1
Para num: 12362


2 es primitivo, luego es primo

In [56]:
def suc_lucas_div(n, p, q, m):
    r = int(n+1)//m # solo porque en este caso el jacobi es -1
    digitos = calcular_digitos(r)
    
    U = [-1, -1]
    U[0], U[1] = 0, 1
    
    U_nuevo = [-1, -1]
    V = -1
    
    V = 2*U[1]-p*U[0]
    
    j = 0
    digit = 0

    while digit < len(digitos):
        
        if digitos[digit] == 0:
            aux = 2*U[0]*U[1]-p*pow(U[0], 2)
            aux = pow(aux, 1, n)
            U_nuevo[0] = aux
            
            aux = pow(U[1], 2)-q*pow(U[0], 2)
            aux = pow(aux, 1, n)
            U_nuevo[1] = aux
            
            V = pow(2*U_nuevo[1]-p*U_nuevo[0],1,n)
            
            if digit >= len(digitos)-2:
                print("\n\n")
                print("U[{}] = {}, \nV[{}] = {}, \n\nU[{}] = {}\n".format(2*j, U_nuevo[0], 2*j, V, 2*j+1, U_nuevo[1]))

            j = 2*j
        else:
            aux = pow(U[1], 2)-q*pow(U[0], 2)
            aux = pow(aux, 1, n)
            U_nuevo[0] = aux
            
            aux = p*pow(U[1],2)-2*q*U[0]*U[1]
            aux = pow(aux, 1, n)
            U_nuevo[1] = aux
            
            V = pow(2*U_nuevo[1]-p*U_nuevo[0],1,n)
            
            if digit >= len(digitos)-2:
                print("\n\n")
                print("U[{}] = {}, \nV[{}] = {}, \n\nU[{}] = {}\n".format(2*j+1, U_nuevo[0], 2*j+1, V, 2*j+2, U_nuevo[1]))

            j = 2*j+1
            
                     
        U[0], U[1] = U_nuevo[0], U_nuevo[1]
        
        digit += 1

In [57]:
for p in factores:
    print(p, r//p)

for elem in factores:
    print("Factor: {}".format(elem))
    suc_lucas_div(n, P, int(Q), elem)
    
suc_lucas_div(n, P, int(Q), 1)

2 18290272972888359279316500480106
2 18290272972888359279316500480106
19 1925291891882985187296473734748
53 690198980108994689785528320004
43117 848401928375738538363824036
210626099398147601381287 173675276
Factor: 2



U[9145136486444179639658250240053] = 2919062146187194170078652929925, 
V[9145136486444179639658250240053] = 32444786707168417748653895320938, 

U[9145136486444179639658250240054] = 35972197399566165238682774605537




U[18290272972888359279316500480106] = 21008262978350532486723653312533, 
V[18290272972888359279316500480106] = 0, 

U[18290272972888359279316500480107] = 28794404462063625522678327136372

Factor: 2



U[9145136486444179639658250240053] = 2919062146187194170078652929925, 
V[9145136486444179639658250240053] = 32444786707168417748653895320938, 

U[9145136486444179639658250240054] = 35972197399566165238682774605537




U[18290272972888359279316500480106] = 21008262978350532486723653312533, 
V[18290272972888359279316500480106] = 0, 

U[182902729728883592793165

Vamos a calcular ahora $U_{\frac{r}{p}}$ con $p\in\{2,19,53,43117,210626099398147601381287\}$.

Sabemos que:

$$U_{\frac{r}{2}} = U_{18290272972888359279316500480106}=21008262978350532486723653312533$$
$$U_{\frac{r}{19}} = U_{1925291891882985187296473734748}=14410717638879893747445724266576$$
$$U_{\frac{r}{53}} = U_{690198980108994689785528320004}=21632150164330732881794537679070$$
$$U_{\frac{r}{43117}} = U_{848401928375738538363824036}=15238186358557607182607231645757$$
$$U_{\frac{r}{210626099398147601381287}} = U_{173675276}=12251700926587395323538401299078$$

Como ninguno de los $U_{\frac{r}{p}}$ con $p$ siendo un factor primo de $r$ sale 0, entonces tenemos que el rango de Lucas $\omega (n)=r$. Luego por un corolario, como el rango de $n$ es $n+1$ tenemos que $n$ es primo.